builtin/logical/pki/cert_util.go

```diff
@@ -1771,3 +1771,26 @@ func parseCertificateFromBytes(certBytes []byte) (*x509.Certificate, error) {
 
 	return x509.ParseCertificate(block.Bytes)
 }
+
+// parseBasicConstraintExtension parses a basic constraint extension, useful if attempting to validate
+// CSRs are requesting CA privileges as Go does not expose it's implementation. Values returned are
+// IsCA, MaxPathLen or error. If MaxPathLen was not set a value of -1 will be returned.
+func parseBasicConstraintExtension(ext pkix.Extension) (bool, int, error) {
+	if !ext.Id.Equal(certutil.ExtensionBasicConstraintsOID) {
+		return false, -1, fmt.Errorf("passed in extension was not a basic constraint extension")
+	}
+
+	// All elements are set to optional here, as it is possible that we receive a CSR with the extension
+	// containing an empty sequence. This is technically allowed and should default to false.
+	type basicConstraints struct {
+		IsCA       bool `asn1:"optional"`
+		MaxPathLen int  `asn1:"optional,default:-1"`
+	}
+	bc := &basicConstraints{}
+	_, err := asn1.Unmarshal(ext.Value, bc)
```

Comment:
If there's additional data here, we want to err too, I believe.

Reply:
Okay, I was trying to be a little more lenient here, kinda like they are within the parsing in [Go](https://github.com/golang/go/blob/master/src/crypto/x509/parser.go#L334-L353) but fair enough.